### PR TITLE
2.x: Add efficient mergeWith(Single|Maybe|Completable) overloads.

### PR DESCRIPTION
**Pull request by akarnokd**
This PR adds specialized overloads to the `mergeWith` operator in `Flowable` and `Observable`.

If accepted, the marbles will be updated in a separate PR.

Related: #5350.

### CODE REVIEW

File: src/main/java/io/reactivex/internal/operators/observable/ObservableMergeWithMaybe.java [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+            if (getAndIncrement() == 0) {
+                queue = null;
+                singleItem = null;
+            }
+        }
+
+        void otherSuccess(T value) {
+            if (compareAndSet(0, 1)) {
+                actual.onNext(value);
+                otherState = 2;
```
Comment (vanniktech) [inline src/main/java/io/reactivex/internal/operators/observable/ObservableMergeWithMaybe.java]: Could we go and use some constants here to improve readability? It's really hard to follow.

File: src/test/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybeTest.java [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+import io.reactivex.plugins.RxJavaPlugins;
+import io.reactivex.processors.PublishProcessor;
+import io.reactivex.subjects.MaybeSubject;
+import io.reactivex.subscribers.TestSubscriber;
+
+public class FlowableMergeWithMaybeTest {
+
+    @Test
+    public void normal() {
+        final TestSubscriber<Integer> ts = new TestSubscriber<Integer>();
```
Comment (vanniktech) [inline src/test/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybeTest.java]: Could be inlined here by using `test()`

File: src/test/java/io/reactivex/internal/operators/observable/ObservableMergeWithCompletableTest.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+import io.reactivex.exceptions.TestException;
+import io.reactivex.functions.Action;
+import io.reactivex.observers.TestObserver;
+import io.reactivex.subjects.*;
+
+public class ObservableMergeWithCompletableTest {
+
+    @Test
+    public void normal() {
+        final TestObserver<Integer> ts = new TestObserver<Integer>();
```
Comment (vanniktech) [inline src/test/java/io/reactivex/internal/operators/observable/ObservableMergeWithCompletableTest.java]: nit: it's an `to` and not `ts`

File: src/test/java/io/reactivex/internal/operators/observable/ObservableMergeWithCompletableTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        )
+        .take(3)
+        .subscribe(ts);
+
+        ts.assertResult(1, 2, 3);
+    }
+
+    @Test
+    public void cancel() {
+        final PublishSubject<Integer> pp = PublishSubject.create();
```
Comment (vanniktech) [inline src/test/java/io/reactivex/internal/operators/observable/ObservableMergeWithCompletableTest.java]: nit: `ps` not `pp`

File: src/test/java/io/reactivex/internal/operators/observable/ObservableMergeWithMaybeTest.java [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+import io.reactivex.functions.Function;
+import io.reactivex.observers.TestObserver;
+import io.reactivex.plugins.RxJavaPlugins;
+import io.reactivex.subjects.*;
+
+public class ObservableMergeWithMaybeTest {
+
+    @Test
+    public void normal() {
+        final TestObserver<Integer> ts = new TestObserver<Integer>();
```
Comment (vanniktech) [inline src/test/java/io/reactivex/internal/operators/observable/ObservableMergeWithMaybeTest.java]: same

File: src/test/java/io/reactivex/internal/operators/observable/ObservableMergeWithSingleTest.java [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+import io.reactivex.functions.Function;
+import io.reactivex.observers.TestObserver;
+import io.reactivex.plugins.RxJavaPlugins;
+import io.reactivex.subjects.*;
+
+public class ObservableMergeWithSingleTest {
+
+    @Test
+    public void normal() {
+        final TestObserver<Integer> ts = new TestObserver<Integer>();
```
Comment (vanniktech) [inline src/test/java/io/reactivex/internal/operators/observable/ObservableMergeWithSingleTest.java]: nit: same

**Review by vanniktech — APPROVED** [review]
I really like the overloads 👍

**Review by vanniktech — APPROVED** [review]

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+                    if (d && empty && otherState == 2) {
+                        queue = null;
+                        actual.onComplete();
+                        return;
+                    }
+                }
+
+                emitted = e;
+                consumed = c;
+                missed = addAndGet(-missed);
```
Comment (davidmoten) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java:320]: I see `requested` not being reduced by emitted amount and `emitted` is now non-volatile field. A bit different to older patterns I am familiar with. Looks handy to eliminate `compareAndSet` calls on `requested` from this spot. Nice.

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+                d.request(prefetch);
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (compareAndSet(0, 1)) {
+                long e = emitted;
+                if (requested.get() != e) {
+                    SimplePlainQueue<T> q = queue;
```
Comment (davidmoten) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java:112]: Are you sure don't need to call `getOrCreateQueue()` here?

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+                    return;
+                }
+            } else {
+                SimplePlainQueue<T> q = getOrCreateQueue();
+                q.offer(t);
+                if (getAndIncrement() != 0) {
+                    return;
+                }
+            }
+            drainLoop();
```
Comment (davidmoten) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java:142]: I'm surprised to see this call `drainLoop()` here rather than `drain` (and in `otherSuccess()`) . Are you sure access to `drainLoop` will be serialized?

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java (hunk 8 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java:112]: Yes. If there is no queue yet or is empty, that means T can be emitted immediately and events from the main source won't be reordered. Otherwise the queue exists and the item should be queued.

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java (hunk 9 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java:142]: When the execution reaches this, the work-in-progress indicator is known to be non-zero as we are in the drain mode already.

Comment (davidmoten) [reply src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java:142]: Ah yes, thanks

**Comment by akarnokd** [comment]
Conflict resolved. There was an expected merge conflict with the `concatWith` PR as they both touched the same test file.

**Review by vanniktech — APPROVED** [review]

**Review by davidmoten — APPROVED** [review]
